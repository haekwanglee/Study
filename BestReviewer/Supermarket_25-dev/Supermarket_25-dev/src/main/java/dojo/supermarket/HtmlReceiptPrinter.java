package dojo.supermarket;

import dojo.supermarket.model.Discount;
import dojo.supermarket.model.Receipt;
import dojo.supermarket.model.ReceiptItem;

public class HtmlReceiptPrinter extends ReceiptPrinter {

    public HtmlReceiptPrinter() {
        this(40);
    }

    public HtmlReceiptPrinter(int columns) {
        super(columns);
    }

    @Override
    String presentReceiptItem(ReceiptItem item) {
        String totalPricePresentation = presentPrice(item.getTotalPrice());
        String name = item.getProduct().getName();

        String line = formatLineWithWhitespace(name, totalPricePresentation);

        if (item.getQuantity() != 1) {
            line += "<tr><td colspan=2>" + presentPrice(item.getPrice()) + " * " + presentQuantity(item) + "</td></tr>\n";
        }
        return line;
    }

    @Override
    String presentDiscount(Discount discount) {
        String name = discount.getDescription() + "(" + discount.getProduct().getName() + ")";
        String value = "-" + presentPrice(discount.getDiscountAmount());

        return formatLineWithWhitespace(name, value);
    }

    @Override
    String presentTotal(Receipt receipt) {
        String name = "<b>Total: ";
        String value = presentPrice(receipt.getTotalPrice());
        return formatLineWithWhitespace(name, value) + "<tr><td colspan=2>&nbsp;</td></tr>\n";
    }

    @Override
    String presentMargin() {
        return "";
    }

    @Override
    String presentHead() {
        return"<table>\n";
    }

    @Override
    String presentEnding() {
        return"</table>";
    }

    @Override
    String formatLineWithWhitespace(String name, String value) {
        StringBuilder line = new StringBuilder();

        line.append("<tr><td>");
        line.append(name);
        line.append("</td><td>");
        line.append(value);
        line.append("</td></tr>\n");
        return line.toString();
    }
}