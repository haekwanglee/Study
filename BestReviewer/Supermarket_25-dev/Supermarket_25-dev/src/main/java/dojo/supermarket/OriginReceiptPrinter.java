package dojo.supermarket;

import dojo.supermarket.model.Discount;
import dojo.supermarket.model.Receipt;
import dojo.supermarket.model.ReceiptItem;

public class OriginReceiptPrinter extends ReceiptPrinter{

    public OriginReceiptPrinter() {
        this(40);
    }

    public OriginReceiptPrinter(int columns) {
        super(columns);
    }

    @Override
    String presentReceiptItem(ReceiptItem item) {
        String totalPricePresentation = presentPrice(item.getTotalPrice());
        String name = item.getProduct().getName();

        String line = formatLineWithWhitespace(name, totalPricePresentation);

        if (item.getQuantity() != 1) {
            line += "  " + presentPrice(item.getPrice()) + " * " + presentQuantity(item) + "\n";
        }
        return line;
    }

    @Override
    String presentDiscount(Discount discount) {
        String name = discount.getDescription() + "(" + discount.getProduct().getName() + ")";
        String value = "-" + presentPrice(discount.getDiscountAmount());

        return formatLineWithWhitespace(name, value);
    }

    @Override
    String presentTotal(Receipt receipt) {
        String name = "Total: ";
        String value = presentPrice(receipt.getTotalPrice());
        return formatLineWithWhitespace(name, value);
    }

    @Override
    String presentMargin() {
        return "\n";
    }

    @Override
    String presentHead() {
        return "";
    }

    @Override
    String presentEnding() {
        return "";
    }

    @Override
    String formatLineWithWhitespace(String name, String value) {
        StringBuilder line = new StringBuilder();
        line.append(name);
        int whitespaceSize = this.columns - name.length() - value.length();
        for (int i = 0; i < whitespaceSize; i++) {
            line.append(" ");
        }
        line.append(value);
        line.append('\n');
        return line.toString();
    }
}
