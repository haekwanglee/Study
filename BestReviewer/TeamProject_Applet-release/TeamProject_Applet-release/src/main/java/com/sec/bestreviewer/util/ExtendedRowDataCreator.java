package com.sec.bestreviewer.util;

import com.sec.bestreviewer.database.Row;
import com.sec.bestreviewer.database.field.EmployeeSchema;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.LinkedList;
import java.util.List;

public class ExtendedRowDataCreator {
    private static final String NAME_DELIMITER = " ";
    private static final String PHONE_NUM_DELIMITER = "-";
    private static final String DATE_PATTERN = "yyyyMMdd";

    private static final int FIRST_NAME_INDEX = 0;
    private static final int LAST_NAME_INDEX = 1;

    private static final int MIDDLE_NUMBER_INDEX = 1;
    private static final int LAST_NUMBER_INDEX = 2;

    public static Row create(Row row) {
        final List<String> updatedDataList = new LinkedList<>(row.getRow());

        appendNames(row, updatedDataList);
        appendPhoneNumbers(row, updatedDataList);
        appendDates(row, updatedDataList);

        return new Row(updatedDataList);
    }

    private static void appendNames(Row row, List<String> extendedList) {
        final String name = row.getValue(EmployeeSchema.NAME.getIndex());
        final String[] splitName = name.split(NAME_DELIMITER);
        extendedList.add(splitName[FIRST_NAME_INDEX]);
        extendedList.add(splitName[LAST_NAME_INDEX]);
    }

    private static void appendPhoneNumbers(Row row, List<String> updatedDataList) {
        final String phoneNumber = row.getValue(EmployeeSchema.PHONE_NUMBER.getIndex());
        final String[] splitPhoneNumber = phoneNumber.split(PHONE_NUM_DELIMITER);
        updatedDataList.add(splitPhoneNumber[MIDDLE_NUMBER_INDEX]);
        updatedDataList.add(splitPhoneNumber[LAST_NUMBER_INDEX]);
    }

    private static void appendDates(Row row, List<String> updatedDataList) {
        final String birthDay = row.getValue(EmployeeSchema.BIRTHDAY.getIndex());
        LocalDate date = LocalDate.parse(birthDay, DateTimeFormatter.ofPattern(DATE_PATTERN));
        updatedDataList.add(String.valueOf(date.getYear()));
        updatedDataList.add(getDateString(date.getMonthValue()));
        updatedDataList.add(getDateString(date.getDayOfMonth()));
    }

    private static String getDateString(int monthOrDay) {
        String date = String.valueOf(monthOrDay);
        return date.length() == 1
                ? "0" + date
                : date;
    }
}
