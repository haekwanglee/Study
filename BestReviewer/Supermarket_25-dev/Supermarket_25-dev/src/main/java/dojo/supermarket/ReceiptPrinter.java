package dojo.supermarket;

import dojo.supermarket.model.Discount;
import dojo.supermarket.model.ProductUnit;
import dojo.supermarket.model.Receipt;
import dojo.supermarket.model.ReceiptItem;

import java.util.Locale;

public abstract class ReceiptPrinter {

    final int columns;
    public ReceiptPrinter(int columns) {
        this.columns = columns;
    }

    public String printReceipt(Receipt receipt) {
        StringBuilder result = new StringBuilder();

        result.append(presentHead());

        for (ReceiptItem item : receipt.getItems()) {
            String receiptItem = presentReceiptItem(item);
            result.append(receiptItem);
        }

        for (Discount discount : receipt.getDiscounts()) {
            String discountPresentation = presentDiscount(discount);
            result.append(discountPresentation);
        }

        result.append(presentMargin());

        result.append(presentTotal(receipt));

        result.append(presentEnding());

        return result.toString();
    }

    abstract String presentReceiptItem(ReceiptItem item);
    abstract String presentDiscount(Discount discount);
    abstract String presentTotal(Receipt receipt);

    abstract String presentMargin();
    abstract String presentHead();
    abstract String presentEnding();
    abstract String formatLineWithWhitespace(String name, String value);

    protected static String presentPrice(double price) {
        return String.format(Locale.UK, "%.2f", price);
    }

    protected static String presentQuantity(ReceiptItem item) {
        return ProductUnit.Each.equals(item.getProduct().getUnit())
                ? String.format("%x", (int)item.getQuantity())
                : String.format(Locale.UK, "%.3f", item.getQuantity());
    }
}
