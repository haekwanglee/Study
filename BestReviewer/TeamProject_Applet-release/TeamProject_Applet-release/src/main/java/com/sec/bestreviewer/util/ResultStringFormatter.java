package com.sec.bestreviewer.util;

import com.sec.bestreviewer.database.Row;
import com.sec.bestreviewer.database.field.EmployeeSchema;

import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public class ResultStringFormatter {
    private static final int YEAR_1990_EMPLOYEE_NUMBER = 90_000000;
    private static final int YEAR_PREFIX_19 = 1900_000000;
    private static final int YEAR_PREFIX_20 = 2000_000000;

    public static List<String> getEmployeeListToFormattedString(List<Row> employeeList) {
        return employeeList.stream()
                .sorted(Comparator.comparing(ResultStringFormatter::getEmployeeNumber))
                .map(ResultStringFormatter::getEmployeeToFormattedString)
                .collect(Collectors.toList());
    }

    private static Integer getEmployeeNumber(Row employee) {
        int employeeNumber = Integer.parseInt(employee.getValue(EmployeeSchema.getPrimaryKeyIndex()));
        return employeeNumber >= YEAR_1990_EMPLOYEE_NUMBER ?
                YEAR_PREFIX_19 + employeeNumber : YEAR_PREFIX_20 + employeeNumber;
    }

    public static String getEmployeeToFormattedString(Row employee) {
        return employee.getValue(EmployeeSchema.EMPLOYEE_NUMBER.getIndex()) + "," +
                employee.getValue(EmployeeSchema.NAME.getIndex()) + "," +
                employee.getValue(EmployeeSchema.CAREER_LEVEL.getIndex()) + "," +
                employee.getValue(EmployeeSchema.PHONE_NUMBER.getIndex()) + "," +
                employee.getValue(EmployeeSchema.BIRTHDAY.getIndex()) + "," +
                employee.getValue(EmployeeSchema.CERTI.getIndex());

    }
}