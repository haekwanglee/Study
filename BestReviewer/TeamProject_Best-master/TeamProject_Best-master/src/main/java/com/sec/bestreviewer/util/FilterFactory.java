package com.sec.bestreviewer.util;

import com.sec.bestreviewer.store.filter.AndFilter;
import com.sec.bestreviewer.store.filter.Filter;
import com.sec.bestreviewer.store.filter.OrFilter;
import com.sec.bestreviewer.store.filter.SingleFilter;

import java.util.List;

import static com.sec.bestreviewer.util.ParamsParser.PairIndex.FIRST_PAIR;
import static com.sec.bestreviewer.util.ParamsParser.PairIndex.SECOND_PAIR;

public class FilterFactory {

    public static Filter buildFilter(List<String> params, OptionParser optionParser) {
        if (optionParser.getAndOption().isAndOn()) {
            Pair<String, String> conditionPair1 = ParamsParser.parser(params, FIRST_PAIR);
            OptionParser optionParser1 = new OptionParser(optionParser.getAndOption().getFirstOptions(), conditionPair1.fieldName);

            Pair<String, String> conditionPair2 = ParamsParser.parser(params, SECOND_PAIR);
            OptionParser optionParser2 = new OptionParser(optionParser.getAndOption().getSecondOptions(), conditionPair2.fieldName);
            return new AndFilter(new SingleFilter(conditionPair1, optionParser1), new SingleFilter(conditionPair2, optionParser2));
        }

        if (optionParser.getOrOption().isOrOn()) {
            Pair<String, String> conditionPair1 = ParamsParser.parser(params, FIRST_PAIR);
            OptionParser optionParser1 = new OptionParser(optionParser.getOrOption().getFirstOptions(), conditionPair1.fieldName);

            Pair<String, String> conditionPair2 = ParamsParser.parser(params, SECOND_PAIR);
            OptionParser optionParser2 = new OptionParser(optionParser.getOrOption().getSecondOptions(), conditionPair2.fieldName);

            return new OrFilter(new SingleFilter(conditionPair1, optionParser1), new SingleFilter(conditionPair2, optionParser2));
        }

        Pair<String, String> conditionPair = ParamsParser.parser(params, FIRST_PAIR);
        return new SingleFilter(conditionPair, optionParser);
    }
}
