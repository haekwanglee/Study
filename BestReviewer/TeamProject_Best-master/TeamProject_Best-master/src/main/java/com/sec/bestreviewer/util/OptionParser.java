package com.sec.bestreviewer.util;

import com.sec.bestreviewer.option.*;

import java.util.Arrays;
import java.util.List;

public class OptionParser {

    public final static List<String> SUPPORTED_OPTIONS = Arrays.asList("-p", "-f", "-l", "-m", "-y", "-d", "-g", "-ge", "-s", "-se", "-a", "-o");
    private final PrintOption printOption;
    private final NameOption nameOption;
    private final BirthDayOption birthDayOption;
    private final PhoneOption phoneOption;
    private final EqualityOption equalityOption;
    private final AndOption andOption;
    private final OrOption orOption;

    public OptionParser(List<String> options, String fieldName) {
        printOption = new PrintOption(options);
        nameOption = new NameOption(options, fieldName);
        birthDayOption = new BirthDayOption(options, fieldName);
        phoneOption = new PhoneOption(options, fieldName);
        equalityOption = new EqualityOption(options);
        andOption = new AndOption(options);
        orOption = new OrOption(options);
    }

    public PrintOption getPrintOption() {
        return printOption;
    }

    public NameOption getNameOption() {
        return nameOption;
    }

    public BirthDayOption getBirthDayOption() {
        return birthDayOption;
    }

    public PhoneOption getPhoneOption() {
        return phoneOption;
    }

    public EqualityOption getEqualityOption() {
        return equalityOption;
    }

    public AndOption getAndOption() {
        return andOption;
    }

    public OrOption getOrOption() {
        return orOption;
    }
}
