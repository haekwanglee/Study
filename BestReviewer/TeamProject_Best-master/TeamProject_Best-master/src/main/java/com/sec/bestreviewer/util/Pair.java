package com.sec.bestreviewer.util;

import java.util.Objects;

public class Pair<F, S> {
    public final F fieldName;
    public final S value;

    private Pair(F fieldName, S value) {
        this.fieldName = fieldName;
        this.value = value;
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Pair)) {
            return false;
        }
        Pair<?, ?> p = (Pair<?, ?>) o;
        return Objects.equals(p.fieldName, fieldName) && Objects.equals(p.value, value);
    }

    @Override
    public int hashCode() {
        return (fieldName == null ? 0 : fieldName.hashCode()) ^ (value == null ? 0 : value.hashCode());
    }

    @Override
    public String toString() {
        return "Pair{" + String.valueOf(fieldName) + " " + String.valueOf(value) + "}";
    }

    public static <A, B> Pair <A, B> create(A a, B b) {
        return new Pair<A, B>(a, b);
    }
}