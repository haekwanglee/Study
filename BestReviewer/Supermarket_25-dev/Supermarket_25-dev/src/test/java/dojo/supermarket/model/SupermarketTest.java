package dojo.supermarket.model;

import dojo.supermarket.HtmlReceiptPrinter;
import dojo.supermarket.OriginReceiptPrinter;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;
import static org.junit.jupiter.params.provider.Arguments.arguments;

public class SupermarketTest {

    // Todo: test all kinds of discounts are applied properly


    @DisplayName("ThreeForTwo discount")
    @ParameterizedTest
    @MethodSource
    public void threeForTwoDiscount(String testProduct,
                                    ProductUnit testUnit,
                                    int testQuantity,
                                    double normalPrice) {
        SupermarketCatalog catalog = new FakeCatalog();
        Product product = new Product(testProduct, testUnit);
        catalog.addProduct(product, normalPrice);

        Teller teller = new Teller(catalog);
        teller.addSpecialOffer(SpecialOfferType.ThreeForTwo, product, 0);

        ShoppingCart cart = new ShoppingCart();
        cart.addItemQuantity(product, testQuantity);

        int actualPayQuantity = testQuantity - (testQuantity / 3);
        double expectedTotalPrice = normalPrice * actualPayQuantity;

        Receipt receipt = teller.checksOutArticlesFrom(cart);

        assertEquals(expectedTotalPrice, receipt.getTotalPrice(), 0.01);
        assertNotEquals(Collections.emptyList(), receipt.getDiscounts());
        assertEquals(1, receipt.getItems().size());
        //Approvals.verify(new ReceiptPrinter().printReceipt(receipt));
    }

    static Stream<Arguments> threeForTwoDiscount() {
        return Stream.of(
                // Buy two toothbrushes, get one free. Normal toothbrush price is €0.99
                arguments("toothbrush", ProductUnit.Each, 3, 0.99),
                arguments("toothbrush", ProductUnit.Each, 7, 0.99)
        );
    }

    @DisplayName("TenPercentDiscount discount")
    @ParameterizedTest
    @MethodSource
    public void tenPercentDiscount(String testProduct,
                                   ProductUnit testUnit,
                                   int testQuantity,
                                   double normalPrice,
                                   int specialPercent) {
        SupermarketCatalog catalog = new FakeCatalog();
        Product product = new Product(testProduct, testUnit);
        catalog.addProduct(product, normalPrice);

        Teller teller = new Teller(catalog);
        teller.addSpecialOffer(SpecialOfferType.TenPercentDiscount, product, specialPercent);

        ShoppingCart cart = new ShoppingCart();
        cart.addItemQuantity(product, testQuantity);

        double actualPayPrice = normalPrice - (normalPrice * specialPercent / 100);
        double expectedTotalPrice = actualPayPrice * testQuantity;

        Receipt receipt = teller.checksOutArticlesFrom(cart);

        assertEquals(expectedTotalPrice, receipt.getTotalPrice(), 0.01);
        assertNotEquals(Collections.emptyList(), receipt.getDiscounts());
        assertEquals(1, receipt.getItems().size());
        //Approvals.verify(new ReceiptPrinter().printReceipt(receipt));
    }

    static Stream<Arguments> tenPercentDiscount() {
        return Stream.of(
                //20% discount on apples, normal price €1.99 per kilo.
                arguments("apples", ProductUnit.Kilo, 3, 1.99, 20),
                //10% discount on rice, normal price €2.49 per bag
                arguments("rice", ProductUnit.Each, 3, 2.49, 10),
                arguments("apples", ProductUnit.Kilo, 5, 1.99, 10)
        );
    }

    @DisplayName("FiveForAmount discount")
    @ParameterizedTest
    @MethodSource
    public void fiveForAmountDiscount(String testProduct,
                                      ProductUnit testUnit,
                                      int testQuantity,
                                      double normalPrice,
                                      double specialPrice) {
        SupermarketCatalog catalog = new FakeCatalog();
        Product product = new Product(testProduct, testUnit);
        catalog.addProduct(product, normalPrice);

        Teller teller = new Teller(catalog);
        teller.addSpecialOffer(SpecialOfferType.FiveForAmount, product, specialPrice);

        ShoppingCart cart = new ShoppingCart();
        cart.addItemQuantity(product, testQuantity);

        int normalPayQuantity = testQuantity % 5;
        int specialPayQuantity = testQuantity / 5;
        double expectedTotalPrice = (normalPayQuantity * normalPrice) + (specialPayQuantity * specialPrice);

        Receipt receipt = teller.checksOutArticlesFrom(cart);

        assertEquals(expectedTotalPrice, receipt.getTotalPrice(), 0.01);
        assertNotEquals(Collections.emptyList(), receipt.getDiscounts());
        assertEquals(1, receipt.getItems().size());
        //Approvals.verify(new ReceiptPrinter().printReceipt(receipt));
    }

    static Stream<Arguments> fiveForAmountDiscount() {
        return Stream.of(
                // Five tubes of toothpaste for €7.49, normal price €1.79
                arguments("toothpaste", ProductUnit.Each, 5, 1.79, 7.49),
                arguments("toothpaste", ProductUnit.Each, 7, 1.79, 7.49)
        );
    }

    @DisplayName("TwoForAmount discount")
    @ParameterizedTest
    @MethodSource
    public void twoForAmountDiscount(String testProduct,
                                     ProductUnit testUnit,
                                     int testQuantity,
                                     double normalPrice,
                                     double specialPrice) {
        SupermarketCatalog catalog = new FakeCatalog();
        Product product = new Product(testProduct, testUnit);
        catalog.addProduct(product, normalPrice);

        Teller teller = new Teller(catalog);
        teller.addSpecialOffer(SpecialOfferType.TwoForAmount, product, specialPrice);

        ShoppingCart cart = new ShoppingCart();
        cart.addItemQuantity(product, testQuantity);

        int normalPayQuantity = testQuantity % 2;
        int specialPayQuantity = testQuantity / 2;
        double expectedTotalPrice = (normalPayQuantity * normalPrice) + (specialPayQuantity * specialPrice);

        Receipt receipt = teller.checksOutArticlesFrom(cart);

        assertEquals(expectedTotalPrice, receipt.getTotalPrice(), 0.01);
        assertNotEquals(Collections.emptyList(), receipt.getDiscounts());
        assertEquals(1, receipt.getItems().size());
        //Approvals.verify(new ReceiptPrinter().printReceipt(receipt));
    }
    static Stream<Arguments> twoForAmountDiscount() {
        return Stream.of(
                //Two boxes of cherry tomatoes for €0.99, normal price €0.69 per box.
                arguments("cherry tomatoes", ProductUnit.Each, 3, 0.69, 0.99),
                arguments("cherry tomatoes", ProductUnit.Each, 7, 0.69, 0.99)
        );
    }

    @Test
    @DisplayName("origin printer : Test using approvals.verify()")
    public void originPrinterGoldenMasterTest() {
        Receipt receipt = makeTestReceipt();
        Approvals.verify(new OriginReceiptPrinter().printReceipt(receipt));
    }

    @Test
    @DisplayName("html printer : Test using approvals.verify()")
    public void htmlPrinterGoldenMasterTest() {
        Receipt receipt = makeTestReceipt();
        Approvals.verify(new HtmlReceiptPrinter().printReceipt(receipt));
    }

    private Receipt makeTestReceipt() {
        List<List<Object>> goldenMasterTestParams = Arrays.asList(
                Arrays.asList("apples", ProductUnit.Kilo, 1.99, 2.5, false),
                Arrays.asList("toothbrush", ProductUnit.Each, 0.99, 5.0, true,
                        SpecialOfferType.TenPercentDiscount, 20.0),
                Arrays.asList("cherry tomato", ProductUnit.Each, 0.69, 2.0, true,
                        SpecialOfferType.TwoForAmount, 0.99),
                Arrays.asList("rice", ProductUnit.Each, 2.49, 4.0, true,
                        SpecialOfferType.TenPercentDiscount, 10.0));

        SupermarketCatalog catalog = new FakeCatalog();
        ShoppingCart cart = new ShoppingCart();
        Teller teller = new Teller(catalog);
        for (List<Object> testParamList: goldenMasterTestParams) {
            TestData testParam = new TestData(testParamList);
            catalog.addProduct(testParam.getProduct(), testParam.getPrice());
            cart.addItemQuantity(testParam.getProduct(), testParam.getQuantity());
            if (testParam.isDiscounted()) {
                teller.addSpecialOffer(testParam.getSpecialOfferType(), testParam.getProduct(), testParam.getArgument());
            }
        }

        return teller.checksOutArticlesFrom(cart);
    }

    static class TestData {
        String productName;
        ProductUnit productUnit;
        Double price;
        Double quantity;
        Boolean isDiscounted;
        SpecialOfferType specialOfferType;
        Double argument;

        Product product;

        public TestData(List<Object> paramObjects) {
            this.productName = (String) paramObjects.get(0);
            this.productUnit = (ProductUnit) paramObjects.get(1);
            this.price = (Double) paramObjects.get(2);
            this.quantity = (Double) paramObjects.get(3);
            this.isDiscounted = (Boolean) paramObjects.get(4);
            if (isDiscounted) {
                this.specialOfferType = (SpecialOfferType) paramObjects.get(5);
                this.argument = (Double) paramObjects.get(6);
            }

            this.product = new Product((String) paramObjects.get(0), (ProductUnit) paramObjects.get(1));
        }

        public Product getProduct() {
            return product;
        }

        public Double getPrice() {
            return price;
        }

        public Double getQuantity() {
            return quantity;
        }

        public Boolean isDiscounted() {
            return isDiscounted;
        }

        public SpecialOfferType getSpecialOfferType() {
            return specialOfferType;
        }

        public Double getArgument() {
            return argument;
        }
    }

}