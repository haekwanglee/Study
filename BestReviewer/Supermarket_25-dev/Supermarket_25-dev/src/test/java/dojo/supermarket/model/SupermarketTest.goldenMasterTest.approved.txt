apples                              4.98
  1.99 * 2.500
toothbrush                          4.95
  0.99 * 5
cherry tomato                       1.38
  0.69 * 2
rice                                9.96
  2.49 * 4
20.0% off(toothbrush)              -0.99
2 for 0.99(cherry tomato)          -0.39
10.0% off(rice)                    -1.00

Total:                             18.89
